### PR TITLE
TRUNK-6731: Add paging support to the obs retrieval API

### PR DESCRIPTION
**Pull request by jwnasambu**
## Description of what I changed

I added paging support to the obs retrieval API by introducing a new getObservations overload that accepts Integer startIndex and Integer maxResults in ObsDAO, ObsService, HibernateObsDAO, and ObsServiceImpl. The existing overloads now delegate to the new paged variant with null, null, maintaining backward compatibility. Paging is applied at the SQL level using JPA setFirstResult/setMaxResults (LIMIT/OFFSET) rather than in memory, and mostRecentN takes precedence over the paging parameters when both are provided. Also added four tests covering page boundaries, a start index beyond the available results, interaction with mostRecentN, and backward compatibility with null paging parameters.

## Issue I worked on

see https://openmrs.atlassian.net/issues?filter=-1&selectedIssue=TRUNK-6731

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `./mvnw clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by dkayiwa** [comment]
@claude review

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java [hunk 1]
```diff
@@ -135,6 +148,13 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
 
 		if (mostRecentN != null && mostRecentN > 0) {
 			query.setMaxResults(mostRecentN);
+		} else {
+			if (startIndex != null && startIndex >= 0) {
+				query.setFirstResult(startIndex);
+			}
+			if (maxResults != null && maxResults > 0) {
+				query.setMaxResults(maxResults);
+			}
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java:168]: We should have some kind of warning if `mostRecentN` and `startIndex` and `maxResults` are set. This should also be noted in the corresponding Javadocs.

File: api/src/main/java/org/openmrs/api/db/ObsDAO.java [hunk 2]
```diff
@@ -100,6 +100,18 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
...
 	        boolean includeVoidedObs, String accessionNumber) throws DAOException;
 
+	/**
+	 * Gets observations with paging support. If mostRecentN is set, the paging parameters are ignored.
+	 *
+	 * @param startIndex the starting index of the result set (optional)
+	 * @param maxResults the maximum number of results to return (optional)
+	 * @since 2.8.0
+	 */
+	public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters, List<Concept> questions,
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/db/ObsDAO.java:118]: 2.8.0 is released, so this can't be from there. Note that the parallel methods are documented on the ObsService which is where this also should be.

File: api/src/main/java/org/openmrs/api/ObsService.java [hunk 3]
```diff
@@ -533,6 +533,19 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
 	        List<Visit> visits, Integer mostRecentN, Integer obsGroupId, Date fromDate, Date toDate,
 	        boolean includeVoidedObs, String accessionNumber) throws APIException;
 
+	/**
+	 * Gets observations with paging support. If mostRecentN is set, the paging parameters are ignored.
+	 *
+	 * @param startIndex the starting index of the result set (optional, 0-based)
+	 * @param maxResults the maximum number of results to return (optional)
+	 * @since 2.8.0
+	 */
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/ObsService.java:570]: Ideally, this would document all of the parameters and not just the newly added ones.

**Review by ibacher — CHANGES_REQUESTED** [review]
Thanks @jwnasambu! A few notes.

**Comment by jwnasambu** [comment]
@claude review

File: api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java [hunk 4]
```diff
@@ -390,15 +392,78 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
...
+		    obsSearchCriteria.getQuestions(), obsSearchCriteria.getAnswers(), obsSearchCriteria.getPersonTypes(),
+		    obsSearchCriteria.getLocations(), sort, obsSearchCriteria.getVisits(), obsSearchCriteria.getMostRecentN(),
+		    obsSearchCriteria.getObsGroupId(), obsSearchCriteria.getFromDate(), obsSearchCriteria.getToDate(),
+		    obsSearchCriteria.getIncludeVoidedObs(), obsSearchCriteria.getAccessionNumber(),
+		    obsSearchCriteria.getStartIndex(), obsSearchCriteria.getMaxResults());
+	}
+
+	/**
+	 * @deprecated as of 3.0.0, use {@link #getObservations(ObsSearchCriteria)}
+	 */
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java:422]: I don't really understand why we're deprecating this? Especially because it's a new method landing in this PR? _Maybe_ we should deprecate the existing method, but not this one.

File: api/src/main/java/org/openmrs/api/ObsService.java [hunk 5]
```diff
@@ -533,6 +534,48 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
...
+	 * @param obsGroupId the Obs.getObsGroupId() to this integer (optional)
+	 * @param fromDate the earliest Obs date to get (optional)
+	 * @param toDate the latest Obs date to get (optional)
+	 * @param includeVoidedObs true/false whether to also include the voided obs (required)
+	 * @param accessionNumber accession number (optional)
+	 * @param startIndex the starting index of the result set (optional, 0-based)
+	 * @param maxResults the maximum number of results to return (optional)
+	 * @since 3.0.0
+	 */
+	@Deprecated(since = "3.0.0")
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/ObsService.java]: As with the `Impl` _we should not be adding new, deprecated API surface_. The _old_ method this replaces should be deprecated.

File: api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java [hunk 6]
```diff
@@ -390,15 +392,78 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
...
 		if (sort == null) {
 			sort = new ArrayList<>();
 		}
 		if (sort.isEmpty()) {
 			sort.add("obsDatetime");
 		}
 
-		return dao.getObservations(whom, encounters, questions, answers, personTypes, locations, sort, visits, mostRecentN,
-		    obsGroupId, fromDate, toDate, includeVoidedObs, accessionNumber);
+		return dao.getObservations(obsSearchCriteria.getWhom(), obsSearchCriteria.getEncounters(),
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java]: Why doesn't this just call the `dao` method that accepts an `obsSearchCriteria`?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java [hunk 7]
```diff
@@ -119,27 +124,68 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
...
+		cq.orderBy(createOrderList(cb, root, obsSearchCriteria.getSort()));
 
 		TypedQuery<Obs> query = session.createQuery(cq);
 
+		Integer mostRecentN = obsSearchCriteria.getMostRecentN();
+		Integer startIndex = obsSearchCriteria.getStartIndex();
+		Integer maxResults = obsSearchCriteria.getMaxResults();
+
 		if (mostRecentN != null && mostRecentN > 0) {
+			if (startIndex != null || (maxResults != null && maxResults > 0)) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java]: We can, similar to how we exclude `maxResults == 0`, also include `startIndex == 0`. Arguably, that's actually the defensible place to skip the warning and we _should_ warn if `maxResults == 0` since this method can return more than `maxResults` results.

File: api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java [hunk 8]
```diff
@@ -390,15 +392,78 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
...
+	/**
+	 * @see org.openmrs.api.ObsService#getObservations(ObsSearchCriteria)
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public List<Obs> getObservations(ObsSearchCriteria obsSearchCriteria) throws APIException {
+		List<String> sort = obsSearchCriteria.getSort();
 		if (sort == null) {
 			sort = new ArrayList<>();
 		}
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java]: We should always ensure that `sort` is a copy.

File: api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java [hunk 9]
```diff
@@ -390,15 +392,78 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
...
+		builder.setSort(sort);
+		builder.setVisits(visits);
+		builder.setMostRecentN(mostRecentN);
+		builder.setObsGroupId(obsGroupId);
+		builder.setFromDate(fromDate);
+		builder.setToDate(toDate);
+		builder.setIncludeVoidedObs(includeVoidedObs);
+		builder.setAccessionNumber(accessionNumber);
+
+		return this.getObservations(builder.createObsSearchCriteria());
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/impl/ObsServiceImpl.java]: It seems like we could've just kept the existing implementation here?

**Review by ibacher — COMMENTED** [review]
Thanks for this! One higher-level comment: I think this doesn't _actually_ have tests directly for the methods with `ObsSearchCriteria` which seems like an oversight?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java (hunk 1 above)
Comment (ibacher) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java]: @jwnasambu Did you see this?

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java [hunk 10]
```diff
@@ -111,35 +119,87 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
 	}
 
 	/**
+	 * @deprecated as of 3.0.0, use {@link #getObservations(ObsSearchCriteria)}
 	 * @see org.openmrs.api.db.ObsDAO#getObservations(List, List, List, List, List, List, List, List,
 	 *      Integer, Integer, Date, Date, boolean, String)
 	 */
 	@Override
+	@Deprecated(since = "3.0.0")
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java]: Why are we deprecating this method? That doesn't seem necessary.

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java [hunk 11]
```diff
@@ -111,35 +119,87 @@ public List<Obs> getObservations(List<Person> whom, List<Encounter> encounters,
 	}
 
 	/**
+	 * @deprecated as of 3.0.0, use {@link #getObservations(ObsSearchCriteria)}
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java]: Deprecation annotations should come _after_ the `@see` annotation that points to the main docs for this.

Comment (ibacher) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernateObsDAO.java]: This applies in multiple places in this PR.

**Review by ibacher — APPROVED** [review]
